### PR TITLE
xcpretty: Support more build steps.

### PR DESCRIPTION
**Pull request by byohay**

### CODE REVIEW

File: features/steps/formatting_steps.rb [hunk 1]
```diff
@@ -19,6 +19,54 @@
...
+Given(/^I have a note$/) do
+  add_run_input SAMPLE_NOTE
+end
+
+Given(/^I have an auxiliary file to write$/) do
+  add_run_input SAMPLE_WRITE_AUXILIARY_FILE
+end
+
+Given(/^I have a file to process$/) do
+  add_run_input SAMPLE_PROCESSING
```
Comment (DekelAvrahami) [inline features/steps/formatting_steps.rb]: ->`SAMPLE_FILE_PROCESSING`

File: features/steps/formatting_steps.rb [hunk 2]
```diff
@@ -391,3 +439,51 @@
 Then(/^I should see the compiling asset catalog message$/) do
   run_output.should start_with("▸ Compiling asset catalog")
 end
+
+Then(/^I should see the creating build directory message$/) do
+  run_output.should start_with("▸ Creating build directory")
+end
+
+Then(/^I should see the mkdir message$/) do
+  run_output.should start_with("▸ Mkdir OnePasswordExtension.framework in target 1PasswordExtension")
```
Comment (DekelAvrahami) [inline features/steps/formatting_steps.rb]: Don't expect the directory name and target.

File: features/steps/formatting_steps.rb [hunk 3]
```diff
@@ -391,3 +439,51 @@
...
+Then(/^I should see the copy PNG file message$/) do
+  run_output.should start_with("▸ Copying PNG file")
+end
+
+Then(/^I should see the copy TIFF file message$/) do
+  run_output.should start_with("▸ Copying TIFF file")
+end
+
+Then(/^I should see the link storyboards message$/) do
+  run_output.should start_with("▸ Linking storyboards for target MobileApp")
```
Comment (DekelAvrahami) [inline features/steps/formatting_steps.rb]: Don't expect the target name.

File: features/steps/formatting_steps.rb [hunk 4]
```diff
@@ -391,3 +439,51 @@
...
+end
+
+Then(/^I should see the note message$/) do
+  run_output.should start_with("▸ Note: detected encoding of input file as Unicode (UTF-8)")
```
Comment (DekelAvrahami) [inline features/steps/formatting_steps.rb]: Don't expect the message itself.

File: features/steps/formatting_steps.rb [hunk 5]
```diff
@@ -391,3 +439,51 @@
...
+end
+
+Then(/^I should see the write auxiliary file message$/) do
+  run_output.should start_with("▸ Writing auxiliary file Specta.hmap for target libSpecta-iOS")
```
Comment (DekelAvrahami) [inline features/steps/formatting_steps.rb]: Don't expect file name and target.

File: features/steps/formatting_steps.rb [hunk 6]
```diff
@@ -391,3 +439,51 @@
...
+end
+
+Then(/^I should see the process file message$/) do
+  run_output.should start_with("▸ Processing /Users/jenkins/.jenkins/workspace/temp/Repo/test_package/Debug-iphoneos/test_packageBundle/de.lproj/Localizable.strings")
```
Comment (DekelAvrahami) [inline features/steps/formatting_steps.rb]: Don't expect file path and name.

File: features/steps/formatting_steps.rb [hunk 7]
```diff
@@ -391,3 +439,51 @@
...
+end
+
+Then(/^I should see the sign identity message$/) do
+  run_output.should start_with("▸ Signing identity \"-\"")
```
Comment (DekelAvrahami) [inline features/steps/formatting_steps.rb]: Don't expect the identity.

File: lib/xcpretty/formatters/formatter.rb [hunk 8]
```diff
@@ -53,6 +53,18 @@ def format_touch(file_path, file_name);                    EMPTY; end
...
+    def format_mkdir(directory_name, target);                  EMPTY; end
+    def format_process_product_packaging(entitlements, app_name); EMPTY; end
+    def format_ditto(source_file, target_file);                EMPTY; end
+    def format_compile_dtrace_script(file_name);               EMPTY; end
+    def format_copy_png_file(source_file, target_file);        EMPTY; end
+    def format_copy_tiff_file(source_file, target_file);       EMPTY; end
+    def format_link_storyboards(target)                        EMPTY; end
+    def format_note(content)                                   EMPTY; end
+    def format_write_auxiliary_file(file_name, target_name)    EMPTY; end
+    def format_processing(file_path)                           EMPTY; end
```
Comment (DekelAvrahami) [inline lib/xcpretty/formatters/formatter.rb]: ->`format_processing_file`

File: lib/xcpretty/parser.rb [hunk 9]
```diff
@@ -217,6 +217,61 @@ module Matchers
     # @regex Captured groups
     WRITE_AUXILIARY_FILES = /^Write auxiliary files/
 
+    # @regex Captured groups
+    # $1 = directory_path
+    # $2 = target
+    CREATE_BUILD_DIRECTORY = /^CreateBuildDirectory\s(\\ |[^ ]+)\s\(in target: (.*)\)/
```
Comment (DekelAvrahami) [inline lib/xcpretty/parser.rb]: 1. add `_MATCHER` suffix. Same everywhere.
2. I think you can replace `\\ |[^ ]+` with `.+`
3. I see that existing matchers capture both full path and name. Do you Want to do it here too? same everywhere.

File: lib/xcpretty/formatters/simple.rb [hunk 10]
```diff
@@ -59,6 +59,60 @@ def format_compile_asset_catalog(file_path)
       format("Compiling asset catalog", file_path)
     end
 
+    def format_create_build_directory(directory_path, target)
+      format(
```
Comment (DekelAvrahami) [inline lib/xcpretty/formatters/simple.rb]: Why not in one line (it is less than 100 chars)? Same in `format_process_product_packaging`.

File: lib/xcpretty/formatters/simple.rb [hunk 11]
```diff
@@ -59,6 +59,60 @@ def format_compile_asset_catalog(file_path)
...
+    def format_copy_png_file(source_file, target_file)
+      format("Copying PNG file", "#{source_file}")
+    end
+
+    def format_copy_tiff_file(source_file, target_file)
+      format("Copying TIFF file", "#{source_file}")
+    end
+
+    def format_link_storyboards(target)
+      format("Linking storyboards for target", "#{target}")
```
Comment (DekelAvrahami) [inline lib/xcpretty/formatters/simple.rb]: let's be consistent by using `in target` everywhere (instead of `for target`). Same below.

File: lib/xcpretty/parser.rb [hunk 12]
```diff
@@ -217,6 +217,61 @@ module Matchers
...
+
+    # @regex Captured groups
+    # $1 = directory_name
+    # $2 = target
+    MKDIR = /^MkDir\s(?:\\ |[^ ])+\/((?:\\ |[^ ])+)\s\(in target: (.*)\)/
+
+    # @regex Captured groups
+    # $1 = entitlements
+    # $2 = app_name
+    PROCESS_PRODUCT_PACKAGING = /^ProcessProductPackaging\s"([^"]*)"\s(?:\\ |[^ ])+\/((?:\\ |[^ ])+)\.xcent/
```
Comment (DekelAvrahami) [inline lib/xcpretty/parser.rb]: I'm not sure what `ProcessProductPackaging` is. 
1. Are you sure it only appears with `.xcent` files?
2. Are you sure `"` is not valid inside the quotes? (with some escaping like `"bal \"bla\" bla"`)
3. You don't need `\\ |[^ ]` unless space is your way to identify the end of the path. In this example you identify the end by `xcent` so it is not needed. Same in all matchers of this PR.

File: lib/xcpretty/parser.rb [hunk 13]
```diff
@@ -217,6 +217,61 @@ module Matchers
...
 
+    # @regex Captured groups
+    # $1 = directory_path
+    # $2 = target
+    CREATE_BUILD_DIRECTORY = /^CreateBuildDirectory\s(\\ |[^ ]+)\s\(in target: (.*)\)/
+
+    # @regex Captured groups
+    # $1 = directory_name
+    # $2 = target
+    MKDIR = /^MkDir\s(?:\\ |[^ ])+\/((?:\\ |[^ ])+)\s\(in target: (.*)\)/
```
Comment (DekelAvrahami) [inline lib/xcpretty/parser.rb]: I think `MkDir\s(?:.*\/)?(.+)\s\(in target: (.*)\)` should do the trick.

File: lib/xcpretty/parser.rb [hunk 14]
```diff
@@ -217,6 +217,61 @@ module Matchers
...
+    # $2 target file
+    COPY_TIFF_FILE = /^CopyTiffFile\s((?:\\ |[^ ])+)\s((?:\\ |[^ ])+)$/
+
+    # @regex Captured groups
+    # $1 target
+    LINK_STORYBOARDS = /^LinkStoryboards\s\(in target: (.*)\)/
+
+    # @regex Captured groups
+    # $1 note content
+    NOTE = /^note: (.*)/
```
Comment (DekelAvrahami) [inline lib/xcpretty/parser.rb]: add `$` at the end.

File: spec/fixtures/constants.rb [hunk 15]
```diff
@@ -586,7 +586,7 @@
...
+    /Applications/Xcode.app/Contents/Developer/usr/bin/actool --output-format human-readable-text --notices --warnings --export-dependency-info /Users/beny/Library/Developer/Xcode/DerivedData/Debug-iphonesimulator/TestHost.build/assetcatalog_dependencies --output-partial-info-plist /Users/beny/Library/Developer/Xcode/DerivedData/sample-etjztiverddwaddrudeteewjzfxw/Build/Intermediates.noindex/sample.build/Debug-iphonesimulator/sample.build/assetcatalog_generated_info.plist --app-icon AppIcon --compress-pngs --enable-on-demand-resources YES --filter-for-device-model iPhone11,8 --filter-for-device-os-version 12.2 --sticker-pack-identifier-prefix com.lightricks.sample.sticker-pack. --target-device iphone --target-device ipad --minimum-deployment-target 10.0 --platform iphonesimulator --product-type com.apple.product-type.application --compile /Users/beny/Library/Developer/Xcode/DerivedData/sample-etjztiverddwaddrudeteewjzfxw/Build/Products/Debug-iphonesimulator/sample.app /Users/beny/developer/video/temp/Sample/Sample/Assets.xcassets
```
Comment (DekelAvrahami) [inline spec/fixtures/constants.rb]: Why did you change this?

File: spec/fixtures/constants.rb [hunk 16]
```diff
@@ -617,6 +617,83 @@
 SAMPLE_SCREENSHOT_FILE = 'RACCommandSpec_enabled_signal_should_send_YES_while_executing_is_YES.png'
 SAMPLE_UNRELATED_IMAGE_FILE = 'apple_raw.png'
 
+SAMPLE_CP_HEADER = %Q(
```
Comment (DekelAvrahami) [inline spec/fixtures/constants.rb:620]: I may be missing something but I don't see where is this used.

File: spec/fixtures/constants.rb [hunk 17]
```diff
@@ -732,3 +809,7 @@
 
 SAMPLE_WILL_NOT_BE_CODE_SIGNED = "FrameworkName will not be code signed because its settings don't specify a development team."
 SAMPLE_FORMAT_OTHER_UNRECOGNIZED_STRING = "This is a string that won't be matched by any other regex."
+
+SAMPLE_GENERIC_WARNING = %Q(
```
Comment (DekelAvrahami) [inline spec/fixtures/constants.rb:813]: I may be missing something but I don't see where is this used.

File: lib/xcpretty/formatters/simple.rb [hunk 18]
```diff
@@ -59,6 +59,60 @@ def format_compile_asset_catalog(file_path)
...
+
+    def format_process_product_packaging(entitlements, app_name)
+      format(
+        "Processing product packaging",
+        "#{entitlements} for app #{app_name}"
+      )
+    end
+
+    def format_ditto(source_file, target_file)
+      format("Ditto", "#{source_file} #{target_file}")
```
Comment (DekelAvrahami) [inline lib/xcpretty/formatters/simple.rb]: Do we want full paths here (the inputs are paths and not names)? Same question everywhere full path is printed.

File: spec/xcpretty/formatters/simple_spec.rb [hunk 19]
```diff
@@ -65,6 +65,86 @@ module XCPretty
...
+
+      it "formats copy PNG file output" do
+        @formatter.format_copy_png_file(
+          "/Users/jenkins/.jenkins/workspace/temp/Repo/test_package/Debug-iphoneos/MyTarget.bundle/IMAGE.PNG",
+          "VideoEngineBundle/IMAGE.PNG"
+        ).should ==
+        "> Copying PNG file /Users/jenkins/.jenkins/workspace/temp/Repo/test_package/Debug-iphoneos/MyTarget.bundle/IMAGE.PNG"
+      end
+
+      it "formats copy PNG file output" do
```
Comment (DekelAvrahami) [inline spec/xcpretty/formatters/simple_spec.rb]: `PNG`->`TIFF`

**Review by DekelAvrahami — CHANGES_REQUESTED** [review]
Round. Generally it looks like this code is relevant for output of the new build system we've switched to recently. However, the rest of the code is written for the old one. For example, `in target` is here because of the new build system. In the old build system there were header lines for the target, and all the lines beneath it was in that target. So I think for this issue we'll need to add `in target <target name>` to all the prints. Maybe it is better to deal with it before adding new matchers for the new build system. Anyway, for the new prints you add in this PR, all have to include the target name.

File: lib/xcpretty/parser.rb (hunk 12 above)
Comment (byohay) [reply lib/xcpretty/parser.rb]: 1. You are right, it can also contain other types.
2. There's a pretty low chance that the entitlement's filename will contain quotes. What do you think?
3. Left as is because `xcent` is not an indicator.

File: lib/xcpretty/parser.rb (hunk 9 above)
Comment (byohay) [reply lib/xcpretty/parser.rb]: 1. Done.
2. Yeah, looks like it.	
3. Not all matchers do it, see `PREPROCESS_MATCHER` and `PBXCP_MATCHER`. My rule of thumb is to capture the full path where there might be multiple files with the same name, and just the file name otherwise.

File: spec/fixtures/constants.rb (hunk 15 above)
Comment (byohay) [reply spec/fixtures/constants.rb]: The word `Foundations` was there.

File: lib/xcpretty/formatters/simple.rb (hunk 18 above)
Comment (byohay) [reply lib/xcpretty/formatters/simple.rb]: I followed the rule of thumb I mentioned above.
In this case the file being "dittoed" is keeping the name, so it doesn't make sense to print only the file name,.

**Comment by byohay** [comment]
@DekelAvrahami round complete.

File: lib/xcpretty/parser.rb [hunk 20]
```diff
@@ -223,58 +223,54 @@ module Matchers
 
     # @regex Captured groups
     # $1 = directory_path
-    # $2 = target
-    CREATE_BUILD_DIRECTORY = /^CreateBuildDirectory\s(\\ |[^ ]+)\s\(in target: (.*)\)/
+    CREATE_BUILD_DIRECTORY_MATCHER = /^CreateBuildDirectory\s(.+)/.freeze
```
Comment (DekelAvrahami) [inline lib/xcpretty/parser.rb]: Add `$` at the end. Same for every matcher that reaches the end of the line.

File: lib/xcpretty/parser.rb [hunk 21]
```diff
@@ -223,58 +223,54 @@ module Matchers
...
 
     # @regex Captured groups
     # $1 = directory_name
-    # $2 = target
-    MKDIR = /^MkDir\s(?:\\ |[^ ])+\/((?:\\ |[^ ])+)\s\(in target: (.*)\)/
+    MKDIR_MATCHER = /^MkDir\s(?:.+\/)?(.+)/.freeze
```
Comment (DekelAvrahami) [inline lib/xcpretty/parser.rb]: 1) I think the first `+` should be replaced with `*` (theoretically the directory could be `/<directory name>`).
2) Why in `CREATE_BUILD_DIRECTORY_MATCHER` you capture the whole path, and here only the directory name?

File: lib/xcpretty/formatters/simple.rb [hunk 22]
```diff
@@ -64,54 +64,63 @@ def format_copy_header_file(source, target, build_target)
       format("Copying", File.basename(source), build_target: build_target)
     end
 
-    def format_create_build_directory(directory_path, target)
+    def format_create_build_directory(directory_path, build_target)
       format(
         "Creating build directory",
-        "#{directory_path} in target #{target}"
+        directory_path,
```
Comment (DekelAvrahami) [inline lib/xcpretty/formatters/simple.rb]: Put in one line if less than 100 chars (same everywhere).

File: lib/xcpretty/parser.rb [hunk 23]
```diff
@@ -223,58 +223,54 @@ module Matchers
...
     # @regex Captured groups
     # $1 source file
     # $2 target file
-    DITTO = /^Ditto\s((?:\\ |[^ ])+)\s((?:\\|[^\s])+)/
+    DITTO_MATCHER = /^Ditto\s((?:\\ |[^ ])+)\s((?:\\|[^\s])+)/.freeze
 
     # @regex Captured groups
     # $1 file_name
-    COMPILE_DTRACE_SCRIPT = /^CompileDTraceScript\s(?:\\ |[^ ])+\/((?:\\ |[^ ])+)$/
+    COMPILE_DTRACE_SCRIPT_MATCHER = /^CompileDTraceScript\s(?:\\ |[^ ])+\/((?:\\ |[^ ])+)$/.freeze
```
Comment (DekelAvrahami) [inline lib/xcpretty/parser.rb]: I think `/^CompileDTraceScript\s.*\/(.+)$/.freeze` should work. No? Same for `WRITE_AUXILIARY_FILE_MATCHER`.

File: lib/xcpretty/parser.rb [hunk 24]
```diff
@@ -223,58 +223,54 @@ module Matchers
...
     # @regex Captured groups
     # $1 auxiliary file name
-    # $2 target name
-    WRITE_AUXILIARY_FILE = /^WriteAuxiliaryFile\s(?:\\ |[^ ])+\/((?:\\ |[^ ])+)\s\(in target: (.*)\)/
+    WRITE_AUXILIARY_FILE_MATCHER = /^WriteAuxiliaryFile\s(?:\\ |[^ ])+\/((?:\\ |[^ ])+)$/.freeze
 
     # @regex Captured groups
     # $1 file path
-    PROCESSING = /^Processing\s((?:\\ |[^\s])+)$/
+    PROCESSING_FILE_MATCHER = /^Processing\s((?:\\ |[^\s])+)$/.freeze
```
Comment (DekelAvrahami) [inline lib/xcpretty/parser.rb]: `(?:\\ |[^\s])+`->`.+`

File: lib/xcpretty/parser.rb [hunk 25]
```diff
@@ -223,58 +223,54 @@ module Matchers
...
-    # $2 = target
-    MKDIR = /^MkDir\s(?:\\ |[^ ])+\/((?:\\ |[^ ])+)\s\(in target: (.*)\)/
+    MKDIR_MATCHER = /^MkDir\s(?:.+\/)?(.+)/.freeze
 
     # @regex Captured groups
     # $1 = entitlements
-    # $2 = app_name
-    PROCESS_PRODUCT_PACKAGING = /^ProcessProductPackaging\s"([^"]*)"\s(?:\\ |[^ ])+\/((?:\\ |[^ ])+)\.xcent/
+    # $2 = package name
+    PROCESS_PRODUCT_PACKAGING_MATCHER = /^ProcessProductPackaging\s"([^"]*)"\s.+\/(.+)$/.freeze
```
Comment (DekelAvrahami) [inline lib/xcpretty/parser.rb]: I suggest `[^"]`->`(?:\\"|[^"])` just to be on the safe side.

**Review by DekelAvrahami — CHANGES_REQUESTED** [review]
Round.

File: spec/fixtures/constants.rb [hunk 26]
```diff
@@ -586,7 +586,7 @@
...
+    /Applications/Xcode.app/Contents/Developer/usr/bin/actool --output- format human-readable-text --notices --warnings --export-dependency-info /Users/beny/Library/Developer/Xcode/DerivedData/Debug-iphonesimulator/TestHost.build/assetcatalog_dependencies --output-partial-info-plist /Users/beny/Library/Developer/Xcode/DerivedData/sample-etjztiverddwaddrudeteewjzfxw/Build/Intermediates.noindex/sample.build/Debug-iphonesimulator/sample.build/assetcatalog_generated_info.plist --app-icon AppIcon --compress-pngs --enable-on-demand-resources YES --filter-for-device-model iPhone11,8 --filter-for-device-os-version 12.2 --sticker-pack-identifier-prefix com.lightricks.sample.sticker-pack. --target-device iphone --target-device ipad --minimum-deployment-target 10.0 --platform iphonesimulator --product-type com.apple.product-type.application --compile /Users/beny/Library/Developer/Xcode/DerivedData/sample-etjztiverddwaddrudeteewjzfxw/Build/Products/Debug-iphonesimulator/sample.app /Users/beny/developer/video/temp/Sample/Sample/Assets.xcassets
```
Comment (DekelAvrahami) [inline spec/fixtures/constants.rb]: `--output- format`->`--output-format`

File: spec/fixtures/constants.rb [hunk 27]
```diff
@@ -662,14 +662,14 @@
 )
 
 SAMPLE_COPY_TIFF_FILE = %Q(
-CopyTiffFile /Users/jenkins/.jenkins/workspace/temp/Repo/test_package/Debug-iphonesimulator/MobileApp.app/PlugIns/MobileAppTests.xctest/ImageCompressorInput.tiff MobileAppTests/Resources/ImageCompressorInput.tiff
+CopyTiffFile /Users/jenkins/.jenkins/workspace/temp/Repo/test_package/Debug-iphonesimulator/MobileApp.app/PlugIns/MobileAppTests.xctest/ (in target: foo)ImageCompressorInput.tiff MobileAppTests/Resources/ImageCompressorInput.tiff
```
Comment (DekelAvrahami) [inline spec/fixtures/constants.rb]: The `in target` looks like it's in the wrong place.

File: lib/xcpretty/parser.rb [hunk 28]
```diff
@@ -223,58 +223,54 @@ module Matchers
...
 
     # @regex Captured groups
     # $1 source file
     # $2 target file
-    DITTO = /^Ditto\s((?:\\ |[^ ])+)\s((?:\\|[^\s])+)/
+    DITTO_MATCHER = /^Ditto\s((?:\\ |[^ ])+)\s((?:\\|[^\s])+)/.freeze
```
Comment (DekelAvrahami) [inline lib/xcpretty/parser.rb]: missing `\s` after the second `?:\\`

**Review by DekelAvrahami — CHANGES_REQUESTED** [review]

File: lib/xcpretty/parser.rb (hunk 21 above)
Comment (byohay) [reply lib/xcpretty/parser.rb]: 1. Done.
2. I thought that the name of the directory in `CreateBuildDirectory` is not indicative of anything (like `Intermediates.noindex`) while in `MkDir` it is (like `EnlightUITests.xctest`). But now I see that there are other non-indicative directories with `MkDir`. So I think that I'll take only the name of the dir and will rely on the build target to give context.

**Comment by byohay** [comment]
@DekelAvrahami round done. I also added `$` to `IN_TARGET_MATCHER` so will be required this to be at the end.

File: lib/xcpretty/parser.rb [hunk 29]
```diff
@@ -223,25 +223,25 @@ module Matchers
...
     # $1 = entitlements
     # $2 = package name
-    PROCESS_PRODUCT_PACKAGING_MATCHER = /^ProcessProductPackaging\s"([^"]*)"\s.+\/(.+)$/.freeze
+    PROCESS_PRODUCT_PACKAGING_MATCHER = /^ProcessProductPackaging\s"((?:\\"|[^"])*)"\s.+\/(.+)$/.freeze
 
     # @regex Captured groups
     # $1 source file
     # $2 target file
-    DITTO_MATCHER = /^Ditto\s((?:\\ |[^ ])+)\s((?:\\|[^\s])+)/.freeze
+    DITTO_MATCHER = /^Ditto\s((?:\\ |[^\s])+)\s((?:\\ |[^\s])+)$/.freeze
```
Comment (DekelAvrahami) [inline lib/xcpretty/parser.rb]: `[^\s]`->`[^\ ]` (twice).

**Review by DekelAvrahami — APPROVED** [review]
Fix squash and merge.
